### ISSUE TITLE
Segmentation method and threshold size

### ISSUE THREAD

**Issue by L-xiaolu** [wgcban/ChangeFormer#23]
Hi, sorry to bother you here. Regarding the model getting the difference map or change probability map of A and B, which method did you use to get their change map? Is it K-means or OSTU? What is the threshold size you set? Where is this part of the code located?

**Comment by wgcban** [comment]
Hi @L-xiaolu 

Usually, we use the threshold of 0.5 to get the change prediction map. 

As you can see, the network outputs 2 channels: one corresponds to change probability and the other corresponds to no-change probability  cp ([here](https://github.com/wgcban/ChangeFormer/blob/b0f2d09f2c457a5886f4808ac8239f20bb9db554/models/ChangeFormer.py#L1285). 
Therefore, if we want to get the binary change prediction map we use the PyTorch command torch.argmax(cp, dim=1) which computes the indices of the maximum value of all elements. This indirectly means that we use the threshold of size 0.5 because we are taking the index (0 or 1) with a higher value. In the code, we do it [here](https://github.com/wgcban/ChangeFormer/blob/b0f2d09f2c457a5886f4808ac8239f20bb9db554/models/trainer.py#L174)

I hope this answers your question.

Best
Chaminda

**Comment by L-xiaolu** [comment]
Thank you for your answer. In addition, when I train the base_transformer_pos_s4_dd8 model, I use multi_scale_train=True, loss=fl. But I get the following error:
![image](https://user-images.githubusercontent.com/77192830/169052575-b24db810-5051-4e9a-ab3e-fe545dac9253.png)
Also, I observed that the value of alpha is weird, but it's not clear where is the problem, I hope you can help me
![image](https://user-images.githubusercontent.com/77192830/169054538-cf476a52-2c27-4673-90eb-c35ada901f3b.png)

**Comment by wgcban** [comment]
Hi @L-xiaolu 
NP. 
As I remember I did not implement multi_scale_train on base_transformer_pos_s4_dd8 model (It is only implemented for Changeformer). 

And the reason for weird alpha is probably you are not normalizing the gt values to 0 and 1 (i think now it is 0-255), that is why it is computing the number of occurrences for all the values from 0 to 255. So normalize the gt by dividing it by 255. 

However, I can include a multi-scale train for base_transformer_pos_s4_dd8 model as well. Please allow me one-two days. I will push an update to the codebase.

Best,
Chaminda,
